trpcultivate_phenotypes/src/Service/TripalCultivatePhenotypesTraitsService.php

```diff
@@ -255,163 +308,286 @@ public function insertTrait($trait, $schema = NULL) {
-      $unit = $this->chado->query($sql, [':id' => $unit_id, ':c_id' => $genus_config['unit']['id']])
-        ->fetchObject();
+    $data_type = $this->chado->query($sql, $args)
+      ->fetchCol();
+
+    if (count($data_type) > 1) {
+      // Unit appears to multiple data types.
+      $this->logger->error(
+        'Error. Failed to retrieve data type for unit : @unit in cv : @cv. Multiple data types found for the same unit.',
+        ['@unit' => $unit, '@cv' => $genus_config['unit']['name']],
+        ['drupal_set_message' => TRUE]
+      );
+      throw new \Exception(t('A multiple data type error occurred while retrieving a unit data type.'));
     }
 
-    return $unit;
+    return ($data_type) ? reset($data_type) : 0;
   }
 
   /**
-   * Get Trait Method Unit data type.
+   * Get trait, method and unit combination.
    *
-   * @param integer $unit_id
-   *   Method unit id number (unit cvterm id).
+   * @param string|int $trait
+   *   A string value is the trait name, whereas an integer value is the trait id number.
+   * @param string|int $method
+   *   A string value is the trait method short name, whereas an integer value is the method id number.
+   * @param string|int $unit
+   *   A string value is the method unit name, whereas an integer value is the unit id number.
    *
-   * @return string
-   *   The data type of the unit, either quantitative or qualitative. False if not set.
+   * @return array
+   *   An associative array where the keys are trait, method and unit and the values
+   *   are the cvterm records for each key.
+   *
+   *   0 if any one of trait, method or unit did not return any record.
+   *
+   * @dependencies
+   *   getTraitAsset(), getMethodUnitDataType().
    */
-  public function getMethodUnitDataType($unit_id) {
-    if ($this->terms['unit_type'] <= 0 && !$unit_id) {
-      // Not configured term and no method id provided.
+  public function getTraitMethodUnitCombo(string|int $trait, string|int $method, string|int $unit) {
+    // Trait.
+    $trait_rec = $this->getTraitAsset($trait, 'trait');
+    if (!$trait_rec) {
+      return 0;
+    }
+
+    // Method.
+    $method_rec = $this->getTraitAsset($method, 'method');
+    if (!$method_rec) {
+      return 0;
+    }
+
+    // Unit.
+    $unit_rec = $this->getTraitAsset($unit, 'unit');
+    if (!$unit_rec) {
       return 0;
     }
 
-    $sql = "SELECT value FROM {1:cvtermprop} WHERE cvterm_id = :c_id AND type_id = :t_id LIMIT 1";
-    $data_type = $this->chado->query($sql, [':c_id' => $unit_id, ':t_id' => $this->terms['unit_type']])
-      ->fetchField();
+    // Append unit data type to the unit data object.
+    $unit_id = $unit_rec->cvterm_id;
+    $unit_data_type = $this->getMethodUnitDataType($unit_id);
+    $unit_rec->{'data_type'} = $unit_data_type;
+
+    return [
+      'trait' => $trait_rec,
+      'method' => $method_rec,
+      'unit'  => $unit_rec
+    ];
+  }
+
+  /**
+   * Get trait asset - trait, method or unit.
```

Comment:
As mentioned above, I think it would be really helpful if this new method were clear that it is returning cvterms. I would switch any mention of trait asset or record in here to say cvterm instead since that is what it is :-)